File: demo/src/test/java/GeminiClientTest.java
import com.google.auth.oauth2.GoogleCredentials;
import com.google.cloud.vertexai.VertexAI;
import com.google.cloud.vertexai.api.GenerateContentResponse;
// import com.google.cloud.vertexai.api.PredictionServiceClient;
import com.google.cloud.vertexai.generativeai.GenerativeModel;
import com.google.cloud.vertexai.generativeai.ResponseHandler;

import java.io.FileInputStream;
// import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.File;

public class GeminiClientTest {
    public static void main(String[] args) throws IOException{
        String projectId = "gen-lang-client-0612946846"; // Google Cloud 專案 ID
        String location = "us-central1"; // Vertex AI 位置
        String modelName = "gemini-1.5-pro";
        String textPrompt =
            "Stock Trading Rules:"+
            "1. Today's price : The last number in each sequence represents today's price. Today's price is excluded from determining whether a value is a peak or trough."+
            "2. Peak : It is the local maximum in the sequence before today, meaning it is greater than its immediate left and right neighbors at a specific position and appears only once at that position."+
            "3. Trough : It is the local minimum in the sequence before today, meaning it is less than its immediate left and right neighbors at a specific position and appears only once at that position."+
            "4. Buy Condition : Buy when today's price (the last number) is strictly greater than the peak(the unique maximum value in the sequence before today)."+
            "5. Sell Condition : Sell when today's price (the last number) is strictly less than the through(the unique minimum value in the sequence before today)."+
            "6. No Action Condition : Take no action when today's price (the last number) is equal to or between the peak and the trough, or if there is no peak or trough."+
            "Instructions:"+
            "Based on the prices preceding today (i.e., the numbers before the last one), determine the action for each period, and indicate the appropriate action."+
            "Please briefly answer the following questions using step-by-step reasoning to show your work and explanation of the all process for finding a peak and a trough(according to the peak and trough rules). Do not answer any other question."+
            "Question :"+
            "Q1. 1, 1, 2, 1, 3, 3, 2, 5"+
            "Q2. 3, 3, 2, 2, 3, 3, 1"+
            "Q3. 1, 2, 3, 2, 5"+
            "Q4. 1, 1, 1, 5"+
            "Q5. 1, 2, 3, 4, 5"+
            "Q6. 1, 2, 3, 4, 5, 5"+
            "Q7. 2, 3, 4, 1"+
            "Q8. 2, 1, 3, 5, 4, 2"+
            "Q9. 2, 2, 2, 3, 2, 1"+
            "Q10. 1.2, 1.0, 1.3, 0.5"+
            "Q11. 1.2, 1, 1.3, 0.5";

        String output = textInput(projectId, location, modelName, textPrompt); // 呼叫 textInput
        System.out.println(output);
    }

    public static String textInput(
        String projectId, String location, String modelName, String textPrompt) throws IOException {
        // 服務帳戶金鑰檔案路徑
        String keyFilePath = "C:\\Users\\evach\\Documents\\research plan\\openAI_test\\demo\\credentials\\gen-lang-client-0612946846-d93ef93d6f82.json"; // 金鑰檔案路徑
        // 建立 GoogleCredentials 並設定範圍
        GoogleCredentials credentials = GoogleCredentials
            .fromStream(new FileInputStream(keyFilePath))
            .createScoped("https://www.googleapis.com/auth/cloud-platform");
        // 驗證金鑰檔案是否存在
        File keyFile = new File(keyFilePath);
        if (keyFile.exists()) {
            System.out.println("檔案存在，路徑正確！");
        } else {
            System.out.println("檔案不存在，請確認路徑：" + keyFile.getAbsolutePath());
        }

        // 計算 API 請求回應時間（開始時間）
        long startTime = System.currentTimeMillis();
        // 初始化 Vertex AI 環境
        try (VertexAI vertexAI = new VertexAI(projectId, location, credentials)) {
            // 初始化 GenerativeModel，指定模型名稱
            GenerativeModel model = new GenerativeModel(modelName, vertexAI);
            // 呼叫 generateContent 方法生成內容
            GenerateContentResponse response = model.generateContent(textPrompt);
            String output = ResponseHandler.getText(response); // 提取回應內容
        // 計算 API 回應時間（結束時間）
        long endTime = System.currentTimeMillis();
        double responseTimeInSeconds = (endTime - startTime) / 1000.0; // 轉換為秒
        System.out.printf("API Response Time: %.2f seconds%n", responseTimeInSeconds);
        
        return output;
    }
  }
}

File: demo/src/test/python/GeminiClientTest.py
import os
import json
import time
import google.generativeai as genai
# from vertexai.generative_models import GenerativeModel
# from google.oauth2 import service_account

def get_API(model_name):
    # 從 config.json 讀取 API Key
    current_working_directory = os.getcwd() 
    config_path = os.path.join(current_working_directory, "demo", "src", "main", "resources", "config.json") # 構建相對路徑

    with open(config_path, "r") as file:
        config = json.load(file)
    api_key = config.get("GEMINI_API_KEY")
    if not api_key:
        raise ValueError("Error: API Key not found in config.json")

    genai.configure(api_key = api_key) # 初始化客戶端
    model = genai.GenerativeModel(model_name)
    return model

def execute_query(model_name, prompt, max_output_tokens, temperature):
    model = get_API(model_name)
    start_time = time.time()
    response = model.generate_content(
        prompt,
        generation_config = genai.GenerationConfig(
            max_output_tokens = max_output_tokens,
            temperature = temperature
        )
    )
    end_time = time.time()
    print(f"Query Time: {end_time - start_time:.2f} seconds")
    print("Total Tokens:", response.usage_metadata.total_token_count)
    return response.text


def main():
    prompt = (
    "Stock Trading Rules:"
    "1. Today's price : The last number in each sequence represents today's price. Today's price is excluded from determining whether a value is a peak or trough."
    "2. Peak : It is the local maximum in the sequence before today, meaning it is greater than its immediate left and right neighbors at a specific position and appears only once at that position."
    "3. Trough : It is the local minimum in the sequence before today, meaning it is less than its immediate left and right neighbors at a specific position and appears only once at that position."
    "4. Buy Condition : Buy when today's price (the last number) is strictly greater than the peak(the unique maximum value in the sequence before today)."
    "5. Sell Condition : Sell when today's price (the last number) is strictly less than the through(the unique minimum value in the sequence before today)."
    "6. No Action Condition : Take no action when today's price (the last number) is equal to or between the peak and the trough, or if there is no peak or trough."
    "Instructions:"
    "Based on the prices preceding today (i.e., the numbers before the last one), determine the action for each period, and indicate the appropriate action."
    "Please briefly answer the following questions using step-by-step reasoning to show your work and explanation of the all process for finding a peak and a trough(according to the peak and trough rules). Do not answer any other question."
    "Question :"
    "Q1. 1, 1, 2, 1, 3, 3, 2, 5"
    "Q2. 3, 3, 2, 2, 3, 3, 1"
    "Q3. 1, 2, 3, 2, 5"
    "Q4. 1, 1, 1, 5"
    "Q5. 1, 2, 3, 4, 5"
    "Q6. 1, 2, 3, 4, 5, 5"
    "Q7. 2, 3, 4, 1"
    "Q8. 2, 1, 3, 5, 4, 2"
    "Q9. 2, 2, 2, 3, 2, 1"
    "Q10. 1.2, 1.0, 1.3, 0.5"
    "Q11. 1.2, 1, 1.3, 0.5")
    result = execute_query("gemini-1.5-pro", prompt, 1000, 0.1)
    print(result)

if __name__ == "__main__":
    main()